### ISSUE TITLE
Repeated extra-valid sync in cluster files

### ISSUE THREAD

**Issue by Selutario** [wazuh/wazuh#10816]
|Wazuh version|Component|
|---|---|
| 4.3 | Wazuh cluster |

## Description

While using the `test_cluster_sync` developed at https://github.com/wazuh/wazuh-qa/pull/2166, we found a strange behavior. The following line can be found multiple times in a row in the `cluster.log` file of worker nodes in our test environments:
```
[Integrity sync] Files to create: 0 | Files to update: 0 | Files to delete: 0 | Files to send: 654
```

If the line is repeated with the same values (for instance, 654 extra-valid files) more than 2 times, we can consider that something happened in the integrity sync, making the master node not add those received files. When it happens, it lasts 4-7 synchronization cycles, and then it seems to work. 

In any case, this needs to be investigated and fixed since repeating the same synchronization multiple times could be a problem in big environments where many files are synced every time.

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

**Comment by Selutario** [comment]
## Issue update

Looking at the logs I have realized that the problem does not happen for a single worker but for all of them at the same time:

![image](https://user-images.githubusercontent.com/23361101/141478078-e07e520a-a4bf-4dbb-93aa-c2ad0e0265fc.png)

As it can be seen in the example above, the same synchronization is repeated 3-5 times in all workers, and the datetime that can be seen in each log is very similar for all of them (it happens at the same time). It means that even if the workers are sending the files that the master needs, it doesn't have those files in the end.

The master has a `Local integrity` task that is repeated every 8 seconds. That task is in charge of calculating the MD5 of all files that should be synced. When this problem is happening, the Local integrity calculates the metadata of the same number of files, even if it received new ones. This line was repeated 11 times:
```
2021/09/29 12:05:40 INFO: [Master] [Local integrity] Finished in 0.692s. Calculated metadata of 44086 files.
```

Strange that this happens a few times and then it works properly again. Some hypotheses that could cause workers to send the same files once and again are:
- The workers do not send the files correctly, so the master actually does not have them when Local integrity is performed. Could be discarded since it would be rare that all workers fail at the same time.
- The master receives the files, but they are deleted immediately. 
- The master correctly receives and processes the files. However, Local Integrity is malfunctioning for some reason and the files are re-requested.

I will try to reproduce the problem.

**Comment by Selutario** [comment]
## Issue update

I have reproduced this error again. While it was happening, I have checked that the content of the `client.keys` was identical (49500 agents) on both the master and a worker node. 

I have also verified that the function [get_files_status()](https://github.com/wazuh/wazuh/blob/a16ab5397d24695bcb8e9154e998d4d1e274c1de/framework/wazuh/core/cluster/cluster.py#L231-L260) was not skipping agents that were inside `/var/ossec/queue/agent-groups`. The function has always obtained the MD5 of all the agents contained in said folder.

Finally, I checked if the workers were actually sending files to the master. They were doing it. What's more, the workers were sending agent-groups of agents that are contained in the `client.keys`. 

After this we can determine that there is no problem in the synchronization of the cluster. Therefore, the problem must be somewhere else. It ​seems to happen when many agents are registered and connected at once, and only with the agents added last.

In the master, `wazuh-modulesd`, `remoted` or another service is in charge of checking if each agent-group created corresponds to an agent of the `client.keys`. It is possible that the last added agents take longer to be processed by that service, which results in their agent-groups being deleted even if the corresponding agent exists in the `client.keys`.

I will open an issue for the core team reporting this behaviour so it can be investigated.

**Comment by Selutario** [comment]
## Issue update

We found an error in the fix we developed at #10866. It turns out that, in some extreme cases, if a lot of extra-valid files need to be synced (in my tests I used 50k), the cluster process would stop working.

The origin of the problem is the query created to check if all the agent-groups extra-valid files correspond to an existing agent in wazuh-db. If the list of extra-valid files is very long, the query (containing each ID of the extra-valid files) can be very long too, even exceeding the 64 KB that are supported by the wazuh-db socket.

A new commit was implemented that, in case more than 7500 extra-valid files need to be synced, multiple smaller queries are performed. It is correctly working now.
